neonvm/controllers/metrics.go

```diff
@@ -0,0 +1,71 @@
+package controllers
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type ReconcilerMetrics struct {
+	failing *prometheus.GaugeVec
+}
+
+func MakeReconcilerMetrics() ReconcilerMetrics {
+	m := ReconcilerMetrics{
+		failing: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "reconcile_failing_objects",
+				Help: "Number of objects that are failing to reconcile for each specific controller",
+			},
+			[]string{"controller"},
+		),
+	}
+	metrics.Registry.MustRegister(m.failing)
+	return m
+}
+
+type wrappedReconciler struct {
+	ControllerName string
+	Reconciler     reconcile.Reconciler
+	Metrics        ReconcilerMetrics
+
+	lock    sync.Mutex
+	failing map[client.ObjectKey]struct{}
+}
+
+// WithMetrics wraps a given Reconciler with metrics capabilities.
+func WithMetrics(reconciler reconcile.Reconciler, rm ReconcilerMetrics, cntrlName string) reconcile.Reconciler {
+	return &wrappedReconciler{
+		Reconciler:     reconciler,
+		Metrics:        rm,
+		ControllerName: cntrlName,
+		lock:           sync.Mutex{},
+		failing:        make(map[client.ObjectKey]struct{}),
+	}
+}
+
+func (d *wrappedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res, err := d.Reconciler.Reconcile(ctx, req)
+
+	// This part is executed sequentially since we acquire a mutex lock. It
+	// should be quite fast since a mutex lock/unlock + 1 memory write takes
+	// ~60ns. Having 8 reconile workers, we wait at most 8*60=480ns to acquire
+	// the lock. I (@shayanh) preferred to go with the simplest implementation
+	// as of now. For a more performant solution, if needed, we can switch to an
+	// async approach.
```

Comment:
fwiw, I think it's more than "1 memory write" — hashmap access plus whatever `WithLabelValues(..).Set(..)` does is probably non-trivial.
But yeah, as long as we're not holding a mutex while doing IO we should be good :)

Reply:
I forgot about `WithLabelValues(..).Set(..)` but it's an `atomic.StoreUint64`.

Reply:
Hm, I appreciate the concern for efficiency. In this case, I'd recommend dropping the exact numbers — the metrics handling isn't trivial¹ and IMO it's not worth it to actually collect exact timings. Either way though, I'd be shocked if manipulating the metrics was a bottleneck, given the number of http requests we make while reconciling :sweat_smile:

---

¹ For example, in the case where the metric already exists, we still end up calling:

- [`(*GaugeVec).GetMetricWithLabelValues()`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/gauge.go#L182-L188)
  - [`(*MetricVec).GetMetricWithLabelValues()`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/vec.go#L201-L208)
    - [`(*MetricVec).hashLabelValues()`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/vec.go#L235-L256) (
      - [`validateLabelValues`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/labels.go#L68-L84) (should be relatively cheap)
      - `(*MetricVec).hashAdd{,Byte}` (internally [`hashAdd`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/fnv.go#L29-L42)) (computes FNV64a hash of the label value(s) - realistically, pretty cheap, but non-zero)
    - [`(*metricMap).getOrCreateMetricWithLabelValues()`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/vec.go#L474-L493) (does RWMutex lock/unlock)
      - [`(*metricMap).getMetricWithHashAndLabelValues()`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/vec.go#L522-L532) (does hashmap lookup)
        - [`findMetricWithlabelValues()`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/vec.go#L550-L559) (iterates through hashmap bucket - should be n=1)
          - [`matchLabelValues()`](https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/vec.go#L574-L593) (somewhat complex control flow, but should be pretty cheap for 1 label value, not too long)

[^examples]: For example, in the case where the metric already exists, we still end up transitively calling `(*GaugeVec).GetMetricWithLabelValues()` → `

Reply:
You're right. The set is the `atomic.StoreUint64`. I already removed the numbers from the comment.